[CodingStandard] Rework anonymous namespace section to cover visibility more broadly (#126775)
llvm/llvm-project: llvm/docs/CodingStandards.rst

.. _static:

Restrict Visibility
^^^^^^^^^^^^^^^^^^^

Functions and variables should have the most restricted visibility possible.

> For class members, that means using appropriate `private`, `protected` or `public`
> keyword to restrict their access. For non-member functions, variables and classes,

Reviewer: ```suggestion
For class members, that means using appropriate `private`, `protected`, or `public`
keyword to restrict their access. For non-member functions, variables, and classes,
```

that means restricting visibility to a single `.cpp` file if it's not referenced
outside that file.

> Visibility of file-scoped non-members variables and functions can be restricted to

Reviewer: ```suggestion
Visibility of file-scope non-member variables and functions can be restricted to
```

> the current translation unit by using either `static` keyword or anonymous namespace.

Reviewer: ```suggestion
the current translation unit by using either the `static` keyword or an anonymous namespace.
```

> Anonymous namespaces are a great language feature that tells the C++ compiler that

Reviewer 2: the `great` here feels a bit like editorializing to me, while I don't disagree it kind of stands out wrt to the rest of the paragraph. I wouldn't bother changing it but for future reference.

contrast, when the function is marked static, you don't need to cross-reference
faraway places in the file to tell that the function is local.

Related to this issue, when non-member functions or variables are defined in a

> `.cpp`` file and need not be referenced outside that file, make them `static`.

Reviewer: I think this is really just clarifying the above paragraph. What if we said:
```
When you are looking at "``runHelper``" in the middle of a large C++ file,
you have no immediate way to tell if this function is local to the file.  In
contrast, when the function is marked static, you don't need to cross-reference
faraway places in the file to tell that the function is local. For this reason, non-
member functions and variables should be marked ``static`` rather than added
to an anonymous namespace.
```

Author: I agree with @joker-eph, again distinguishing between *what* needs to be internalized (last paragraph that @joker-eph suggested) vs *how* to achieve that (static for non-member functions and variables) and anonymous namespaces for classes and structs. I'd reword the last paragraph a bit as follows, with some extra/redundant clarification:

> We aim to make functions, variables, and classes as private as possible: If a non-member function or variable is defined in a `.cpp` file and need not be referenced outside that file, it should be made private to this file (by using `static`). Similarly, if a class or struct is defined in a `.cpp` file and need not be referenced outside, it should be made private to this file (by using anonymous namespace).

WDYT?

Author: Or to belabor on this more, may be this should be the first paragraph, saying that things should be as private as possible, and then going into how to achieve that via static or anon namespaces. The section can be renamed to something like *privatize aggressively* or keep it as is.

Reviewer: > Or to belabor on this more, may be this should be the first paragraph, saying that things should be as private as possible, and then going into how to achieve that via static or anon namespaces. The section can be renamed to something like privatize aggressively or keep it as is.

This format would make more sense to me. Basically, "Here's the principle we want to follow and why. Here's how to follow that principle." Should we also strongly encourage using `private` and `protected` access specifiers within classes too? Then the section is basically about all the ways in which you can restrict access to identifiers. (I don't know if I'm scope-creeping too much though.)

Author: Finally got around to updating this based on the suggestion above. @AaronBallman and @joker-eph PTAL.
